tests/src/cli.js

```diff
@@ -52,7 +64,8 @@ describe('CLI regression tests', function () {
     let eslint;
     let cli;
     beforeEach(function () {
-      if (semver.satisfies(eslintPkg.version, '< 6')) {
+      // TODO: it might be possible to get this working on v9 with a lot of wrangling of eslint-plugin-json v2/4
+      if (semver.satisfies(eslintPkg.version, '< 6 || >= 9')) {
```

Comment:
I ended up skipping this because it doesn't feel super critical vs having v9 support out the door, and I'm not sure how relevant or doable it is with flat config.

The related issue seems to be about when you're linting a project that has just JSON files which inherently is now impossible with flat config since it only supports JavaScript based configs; in addition, `eslint-plugin-json` has to be on v4 for flat config, but then that causes issues because the existing `.eslintrc.json` still gets read because of how `no-unused-modules` works and relating to https://github.com/eslint/eslint/issues/18087

```
     Error: Error while loading rule 'rule-to-test/no-unused-modules': ESLint configuration in tests/files/just-json-files/.eslintrc.json#overrides[0] » plugin:json/recommended is invalid:
        - Unexpected top-level property "files".

Referenced from: /home/jones/workspace/projects-oss/eslint-plugin-import/tests/files/just-json-files/.eslintrc.json
```

(this happens even when I introduced a whole new "just-json-files-flat" directory, I believe because `no-unused-modules` wants to slurp up the whole codebase, or something along those lines...)

I feel like overall there is an improvement to be made to `no-unused-modules` in regards to v9 support, but from what I've read it doesn't sound like an immediate blocker and also that there's a _lot_ more potential work needing to be done.

Reply:
Out of curiosity, what was the state of this file when you were seeing that error? I'd like to explore that a bit locally, unless you have something higher priority that you'd like me to look at.

Reply:
I was able to get the test to pass with this flat config (note: the test passes in the import plugin, which is why it's not defined in the config).  I believe this is roughly equivalent to the eslintrc config that was in there already.

```js
import jsonPlugin from 'eslint-plugin-json';

export default [
  {
    files: ['tests/files/just-json-files/*.json'],
    plugins:{
      json: jsonPlugin,
    },
    processor: 'json/json',
    rules: {
      'import/no-unused-modules': [
        'error',
        {
          'missingExports': false,
          'unusedExports': true,
        }
      ],
      ...jsonPlugin.configs.recommended.rules
    },
  },
];
```

Reply:
that doesn't work for me:

![image](https://github.com/user-attachments/assets/40fef578-5cc1-40d5-a107-514182f93320)

<details>
<summary>this is my patch</summary>

```
Index: tests/files/just-json-files/eslint.config.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/tests/files/just-json-files/eslint.config.js b/tests/files/just-json-files/eslint.config.js
new file mode 100644
--- /dev/null	(date 1726706360463)
+++ b/tests/files/just-json-files/eslint.config.js	(date 1726706360463)
@@ -0,0 +1,21 @@
+import jsonPlugin from 'eslint-plugin-json';
+
+export default [
+  {
+    files: ['tests/files/just-json-files/*.json'],
+    plugins:{
+      json: jsonPlugin,
+    },
+    processor: 'json/json',
+    rules: {
+      'import/no-unused-modules': [
+        'error',
+        {
+          'missingExports': false,
+          'unusedExports': true,
+        }
+      ],
+      ...jsonPlugin.configs.recommended.rules
+    },
+  },
+];
Index: tests/src/cli.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/tests/src/cli.js b/tests/src/cli.js
--- a/tests/src/cli.js	(revision 23260f08240b64c370ab3aae7c137331aa379a6b)
+++ b/tests/src/cli.js	(date 1726706413680)
@@ -64,18 +64,24 @@
     let eslint;
     let cli;
     beforeEach(function () {
-      // TODO: it might be possible to get this working on v9 with a lot of wrangling of eslint-plugin-json v2/4
-      if (semver.satisfies(eslintPkg.version, '< 6 || >= 9')) {
+      if (semver.satisfies(eslintPkg.version, '< 6')) {
         this.skip();
       } else {
         if (ESLint) {
-          eslint = new ESLint({
-            useEslintrc: false,
-            overrideConfigFile: './tests/files/just-json-files/.eslintrc.json',
-            rulePaths: ['./src/rules'],
-            ignore: false,
-            plugins: { 'eslint-plugin-import': importPlugin },
-          });
+          if (semver.satisfies(eslintPkg.version, '>= 9')) {
+            eslint = new ESLint({
+              overrideConfigFile: './tests/files/just-json-files/eslint.config.js',
+              plugins: { 'eslint-plugin-import': importPlugin },
+            });
+          } else {
+            eslint = new ESLint({
+              useEslintrc: false,
+              overrideConfigFile: './tests/files/just-json-files/.eslintrc.json',
+              rulePaths: ['./src/rules'],
+              ignore: false,
+              plugins: { 'eslint-plugin-import': importPlugin },
+            });
+          }
         } else {
           cli = new CLIEngine({
             useEslintrc: false,
```
</details>

(also I'm still pretty sure adding a js file to that directory invalids the test because it's specifically meant to be there are just json files 😅)

Reply:
<details>
<summary>as to your original question re my local state for my original comment, here's the patch for that<summary>

```
Index: package.json
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/package.json b/package.json
--- a/package.json	(revision 23260f08240b64c370ab3aae7c137331aa379a6b)
+++ b/package.json	(date 1726706778134)
@@ -85,7 +85,7 @@
     "eslint-module-utils": "file:./utils",
     "eslint-plugin-eslint-plugin": "^2.3.0",
     "eslint-plugin-import": "2.x",
-    "eslint-plugin-json": "^2.1.2",
+    "eslint-plugin-json": "^4.0.0",
     "find-babel-config": "=1.2.0",
     "fs-copy-file-sync": "^1.1.1",
     "glob": "^7.2.3",
Index: tests/src/cli.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/tests/src/cli.js b/tests/src/cli.js
--- a/tests/src/cli.js	(revision 23260f08240b64c370ab3aae7c137331aa379a6b)
+++ b/tests/src/cli.js	(date 1726706849959)
@@ -65,10 +65,16 @@
     let cli;
     beforeEach(function () {
       // TODO: it might be possible to get this working on v9 with a lot of wrangling of eslint-plugin-json v2/4
-      if (semver.satisfies(eslintPkg.version, '< 6 || >= 9')) {
+      if (semver.satisfies(eslintPkg.version, '< 6')) {
         this.skip();
       } else {
         if (ESLint) {
+          if (semver.satisfies(eslintPkg.version, '>= 9')) {
+            eslint = new ESLint({
+              overrideConfigFile: './tests/files/just-json-files-linted/eslint.config.js',
+              plugins: { 'eslint-plugin-import': importPlugin },
+            });
+          } else {
           eslint = new ESLint({
             useEslintrc: false,
             overrideConfigFile: './tests/files/just-json-files/.eslintrc.json',
@@ -76,6 +82,7 @@
             ignore: false,
             plugins: { 'eslint-plugin-import': importPlugin },
           });
+          }
         } else {
           cli = new CLIEngine({
             useEslintrc: false,
@@ -89,7 +96,9 @@
     });
 
     it('throws an error on invalid JSON', () => {
-      const invalidJSON = './tests/files/just-json-files/invalid.json';
+      const invalidJSON = semver.satisfies(eslintPkg.version, '>= 9')
+        ? './tests/files/just-json-files-linted/invalid.json'
+        : './tests/files/just-json-files/invalid.json';
       if (eslint) {
         return eslint.lintFiles([invalidJSON]).then((results) => {
           expect(results).to.eql(
Index: tests/files/just-json-files-linted/eslint.config.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/tests/files/just-json-files-linted/eslint.config.js b/tests/files/just-json-files-linted/eslint.config.js
new file mode 100644
--- /dev/null	(date 1726706769053)
+++ b/tests/files/just-json-files-linted/eslint.config.js	(date 1726706769053)
@@ -0,0 +1,25 @@
+const jsonPlugin = require('eslint-plugin-json');
+const importPlugin = require('../../../src/index');
+
+module.exports = [
+  {
+    files: ['**/*.json'],
+    plugins: {
+      import: importPlugin,
+      json: jsonPlugin,
+    },
+    rules: {
+      'import/no-unused-modules': [
+        'error',
+        {
+          missingExports: true,
+          unusedExports: true,
+        },
+      ],
+    },
+  },
+  {
+    files: ['**/*.json'],
+    ...jsonPlugin.configs.recommended,
+  }
+]
Index: tests/files/just-json-files-linted/invalid.json
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/tests/files/just-json-files-linted/invalid.json b/tests/files/just-json-files-linted/invalid.json
new file mode 100644
--- /dev/null	(date 1726706769093)
+++ b/tests/files/just-json-files-linted/invalid.json	(date 1726706769093)
@@ -0,0 +1,1 @@
+,
```
</details>

Reply:
oops, updated my reply without refreshing and seeing your reply.  Here's the config and test setup.  (I forgot to past in the version with the patched processor of the old json plugin)

```js
import jsonPlugin from 'eslint-plugin-json';

if (!jsonPlugin.processors.json) {
  jsonPlugin.processors.json = jsonPlugin.processors['.json'];
}

export default [
  {
    files: ['tests/files/just-json-files/*.json'],
    plugins:{
      json: jsonPlugin,
    },
    processor: 'json/json',
    rules: {
      'import/no-unused-modules': [
        'error',
        {
          'missingExports': false,
          'unusedExports': true,
        }
      ],
      ...jsonPlugin.configs.recommended.rules
    },
  },
];
```

This is the test set up
```js
if (semver.satisfies(eslintPkg.version, '>= 9')) {
  eslint = new ESLint({
    overrideConfigFile: './tests/files/just-json-files/eslint.config.mjs',
    plugins: { 'eslint-plugin-import': importPlugin },
  });
} else {
  eslint = new ESLint({
    useEslintrc: false,
    overrideConfigFile: './tests/files/just-json-files/.eslintrc.json',
    rulePaths: ['./src/rules'],
    ignore: false,
    plugins: { 'eslint-plugin-import': importPlugin },
  });
}
```

But yours looks the same, minus the processor patch.  And if yours was failing due to the processor patch, it would be a different error.  I was seeing the error you're getting when I didn't have the `files` in the config set to be relative to the root of the repo.  But once I realized that issue, I didn't see that anymore.   The only other minor difference i see is that I used an `mjs` extension, but that shouldn't matter.

Reply:
awesome sweet that's passing for me locally now too - have pushed that up